File: route_config.go
package router

import (
	"net/http"
	"strings"
)

var httpMethods = [9]string{"GET", "POST", "PUT", "DELETE", "HEAD", "CONNECT", "OPTIONS", "TRACE", "PATCH"}

type RouteConfig struct {
	route               string
	handlerfn           http.HandlerFunc
	methods             []string
	acceptTrailingSlash bool
	allowCors           bool
}

// Accept a trailing slash a single route if client added one to the request even if you configured none in your path.
func (rc *RouteConfig) AcceptTrailingSlash(b bool) {
	rc.acceptTrailingSlash = b
}

// Set allowed methods for a single route.
func (rc *RouteConfig) Methods(methods ...string) {
	rc.methods = methods
}

// Set headers to allow CORS requests for a single route. Default is false.
func (rc *RouteConfig) AllowCORS(b bool) {
	rc.allowCors = b
}

// checking functions - not public

func (rt *RouteConfig) checkMethods(r *http.Request) bool {
	if len(rt.methods) == 0 {
		return true
	}
	for _, m := range rt.methods {
		if strings.EqualFold(m, r.Method) {
			return true
		}
	}
	return false
}

func (rt *RouteConfig) checkAllowCors(w http.ResponseWriter) {
	if !rt.allowCors {
		return
	} else {
		if len(rt.methods) == 0 {
			rt.methods = httpMethods[:]
		}
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Access-Control-Allow-Methods", strings.Join(rt.methods, ", "))
		w.Header().Set("Accept", "*/*")
	}
}

File: src/route_config.rs
use crate::*;
#[derive(Clone)]
pub struct RouteConfig {
    pub (in super) route: &'static str,
    pub (in super) handlerfn: fn(req: Request<Body>) -> Response<Body>,
    pub (in super) methods: Vec<&'static str>,
    pub (in super) accept_trailing_slash: bool,
    pub (in super) allow_cors: bool,
}

impl RouteConfig {
    pub (in crate) fn check_methods(&self, req: &Request<Body>) -> bool {
        if self.methods.len() == 0 {
            return true
        }
        for m in &self.methods {
            if m.to_lowercase() == req.method().as_str().to_lowercase() {
                return true
            }
        }
        return false
    }
}
